### PR TITLE
implemented sign agreement

### PR DESCRIPTION
**Pull request by Robka99**

### CODE REVIEW

File: canisters/src/vts/src/lib.rs [hunk 1]
```diff
@@ -143,5 +143,29 @@ fn create_agreement(
...
 }
 
+#[ic_cdk::update]
+fn sign_agreement(vh_provider_public_key: String) -> VTSResult<()> {
+    let caller = ic_cdk::api::caller();
+    ic_cdk::println!("requested agreement signing by {}", caller);
+    AGREEMENTS.with(|agreements| {
+        let agreements = agreements.borrow_mut();
+
+        let agreement_key = agreements.iter().find(|(_, agreement)| agreement.vh_provider.to_text() == vh_provider_public_key);
```
Comment (lavrd) [inline canisters/src/vts/src/lib.rs]: Why do we need to iterate over whole map? Key in this map is vehicle customer provider public key. Can we just use get method?

File: canisters/src/vts/src/lib.rs [hunk 2]
```diff
@@ -143,5 +143,29 @@ fn create_agreement(
...
+
+        match agreement_key {
+            Some((_, mut agreement)) => {
+                if agreement.vh_customer != caller {
+                    return Err(Error::Internal);
```
Comment (lavrd) [inline canisters/src/vts/src/lib.rs]: Create new error variant to describe such error.

File: canisters/src/vts/src/lib.rs [hunk 3]
```diff
@@ -143,5 +143,29 @@ fn create_agreement(
...
+                }
+
+                agreement.state = AgreementState::Signed;
+
+                Ok(())
+            },
+            None => Err(Error::Internal),
```
Comment (lavrd) [inline canisters/src/vts/src/lib.rs]: If we didn't find agreement return new error variant, something like NotFound.

**Review by lavrd — CHANGES_REQUESTED** [review]

File: canisters/src/vts/src/lib.rs (hunk 1 above)
Comment (Robka99) [reply canisters/src/vts/src/lib.rs]: I had some problems with the get function.
Error messages such as:
`binary operation == cannot be applied to type AgreementState`
or:
```
mismatched types
expected reference &u128
   found reference &std::string::String
```

Maybe you have some ideas?

Comment (lavrd) [reply canisters/src/vts/src/lib.rs]: You can pass agreement id as a parameter to use it as a key. I forgot that key is not provider public key.

File: canisters/src/vts/tests/agreement.rs [hunk 4]
```diff
@@ -20,11 +20,20 @@ async fn test_create_agreement() {
...
+    let daily_usage_fee = "100".to_string();
+    let gas_price = "10".to_string();
+
+    let agreement_id = create_agreement(&agent, &canister_id, &name, &vh_customer, &daily_usage_fee, &gas_price).await.unwrap();
 
-    // let result =
-    //     create_agreement(&agent, &canister_id, &name, &vh_customer, &daily_usage_fee, &gas_price)
-    //         .await;
-    // assert!(result.is_err(), "Should return error on duplicate agreement");
+    let result = sign_agreement(&agent,&canister_id, &agreement_id).await;
```
Comment (lavrd) [inline canisters/src/vts/tests/agreement.rs]: Can you execute `make lint`, please? It looks like your source code is not formatted. We will do CI/CD later for it.

**Review by lavrd — CHANGES_REQUESTED** [review]

File: canisters/src/vts/tests/agreement.rs [hunk 5]
```diff
@@ -20,11 +20,23 @@ async fn test_create_agreement() {
...
+            .await
+            .unwrap();
 
-    // let result =
-    //     create_agreement(&agent, &canister_id, &name, &vh_customer, &daily_usage_fee, &gas_price)
-    //         .await;
-    // assert!(result.is_err(), "Should return error on duplicate agreement");
+    let result = sign_agreement(&agent, &canister_id, &agreement_id).await;
+    assert!(result.is_ok(), "Should successfully sign the agreement");
 }
```
Comment (lavrd) [inline canisters/src/vts/tests/agreement.rs:41]: Please, add more tests. Like try to sign an agreement twice. Try to sign agreement which is not created. Check for particular errors.

**Review by lavrd — CHANGES_REQUESTED** [review]

File: canisters/src/vts/tests/agreement.rs (hunk 5 above)
Comment (Robka99) [reply canisters/src/vts/tests/agreement.rs:41]: I have now added two more tests:
An ID that does not exist and sign agreement twice.

File: canisters/src/vts/tests/agreement.rs [hunk 6]
```diff
@@ -20,11 +20,52 @@ async fn test_create_agreement() {
...
+
+    let result_first = sign_agreement(&agent, &canister_id, &agreement_id).await;
+    assert!(result_first.is_ok(), "Should successfully sign the agreement the first time");
 
-    // let result =
-    //     create_agreement(&agent, &canister_id, &name, &vh_customer, &daily_usage_fee, &gas_price)
-    //         .await;
-    // assert!(result.is_err(), "Should return error on duplicate agreement");
+    let result_second = sign_agreement(&agent, &canister_id, &agreement_id).await;
+    assert!(result_second.is_err(), "Should fail to sign the agreement a second time");
```
Comment (lavrd) [inline canisters/src/vts/tests/agreement.rs]: Check for particular error to be sure that we receive Error::AlreadyExists

File: canisters/src/vts/tests/agreement.rs [hunk 7]
```diff
@@ -20,11 +20,52 @@ async fn test_create_agreement() {
...
+    assert!(result.is_ok(), "Should successfully sign the agreement");
+}
+
+#[tokio::test]
+async fn test_sign_nonexistent_agreement() {
+    let (agent, canister_id) = init_agent().await;
+    let nonexistent_agreement_id = 999999; // An ID that doesn't exist.
+
+    let result = sign_agreement(&agent, &canister_id, &nonexistent_agreement_id).await;
+    assert!(result.is_err(), "Should fail to sign a nonexistent agreement");
```
Comment (lavrd) [inline canisters/src/vts/tests/agreement.rs]: Check for Error::NotFound

**Review by lavrd — CHANGES_REQUESTED** [review]

**Review by lavrd — APPROVED** [review]
Well done!

**Pull request merged**